Package.swift

```diff
@@ -87,6 +87,7 @@ var targets: [PackageDescription.Target] = [
 
 let package = Package(
     name: "swift-nio",
+    platforms: [.iOS(.v10), .macOS(.v10_12), .tvOS(.v10), .watchOS(.v3)],
```

Comment:
`os_unfair_lock` is only supported on these platforms and above. They are all _very_ old, so support shouldn't a problem, rolling this in with the minimum Swift version changes is probably a good idea. Neither break SemVer, but have the potential to cause pain.

Reply:
Agreed they're very old and unlikely for users to _actually_ be using them. 

However, the actual problem is that adding platforms _does_ cause a _lot_ of pain: we will break any dependent which doesn't already specify platforms. In the past we added platforms but quickly removed them because of the pain. See also: https://github.com/apple/swift-nio/issues/1049

Reply:
We definitely can't do this, this is SemVer major. @Davidde94 you can just `#if` use `os_unfair_lock` when available.

Reply:
I don't follow why this would be major, but changing the minimum language version isn't?

Reply:
Also I don't think you can get around this using `#if`. You can you `@available` but that would become a mess IMO

Reply:
because SwiftPM will still resolve a version with these guards and then complain that the depending project doesn't have matching OS versions. So the build will fail.

For changing the minimum language versions, older SwiftPMs will just ignore all new versions and correctly resolve the old ones.

Reply:
David is right that `#if` is not sufficient to determine what lock to use: you have to do runtime version checking.

I don't know what the performance implications are for Darwin of having to decide what lock implementation we're using at runtime, but I imagine they may overwhelm the gain of using the unfair lock.